Add Application api stubs

pbn4 opened the pull request:
Closes https://github.com/bloom-housing/bloom/issues/130 and https://github.com/bloom-housing/bloom/issues/140

@bk3c I decided to move swagger + listings service stuff here since there were conflicts and the swagger part overlaps. It was easier for me to simply copy the listings changes here. 

@bk3c @bencpeters @software-project @jaredcwhite I renamed files containing entities to `className.entity.ts` so that swagger plugin can pick them up by suffix. Otherwise we would have to define DTOs fo every entity class manually with all the properties basically copied from the original entity. 

[\*] Make Listings service use swagger plugin
[\*] Implement `/applications` route 
[\*] Implement `/user/{userId}/applications`
[\*] Write e2e tests for `/user/{userId}`
[] Write e2e tests for `/applications` (not doing that since roles e.g. 'admin' are missing, route is protected by an auth guard requiring JWT token to contain roles prop)

At the moment API accepts arbitrary json as application (see application prop in POST/PUT body dto)

jaredcwhite commented on backend/core/src/application/application.controller.ts:
```diff
@@ -0,0 +1,18 @@
+import { Application } from "../entity/application.entity"
+
+@Controller("application")
+export class ApplicationController {
```
For consistency (and also because Application can have many meanings), can we rename this to `ApplicationsController` (plural) like we have `ListingsController` (and same for the service)?

pbn4 replied:
Done

jaredcwhite commented on backend/core/src/application/application.service.ts:
```diff
@@ -0,0 +1,14 @@
+@Injectable()
+export class ApplicationService {
+  constructor(
+    @InjectRepository(Application) private readonly applicationRepo: Repository<Application>
```
Can we use the same ActiveRecord-style pattern here as we do for the Listings?

pbn4 replied:
Done. @bencpeters I guess the same applies to user and auth modules?

bencpeters replied:
Yes

pbn4 commented:
@bencpeters Should I create a module for Applications the same way you did for User and Auth?

pbn4 commented:
Another question. What's the problem here?

```
// Use require because of the CommonJS/AMD style export.
// See https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require
import dbOptions = require("../../ormconfig.test")
```

 @bencpeters Are you the author? Is this related to our monorepo setup? I have a side project using NestJS and I did not hit that anywhere.

pbn4 commented:
@software-project I have put first test for applications in test/applications, is that correct? 

@jaredcwhite @bencpeters @bk3c @software-project  Regarding our discussion about having a test DB I made an initial step in this direction [here](https://github.com/bloom-housing/bloom/pull/357/files#diff-b6549d14994f69cfcba779f356a42529). I created `ormconfig.test.ts` a separate config for the test DB and added `clearDb` a utility dropping every table so the test DB is clear. In my test I call it in `beforeAll` since there is a single test at the moment. Normally I believe it should be called in `beforeEach` so that our tests are independent from each other.

pbn4 commented:
I recognize that we had a discussion about having permission guard at service level. According to [this](https://docs.nestjs.com/faq/request-lifecycle) services can not be decorated with guards decorators. 

So far I added auth guards a at controller level: 

```
...
@Controller("user")
@UseGuards(JwtAuthGuard)
@UseGuards(OwnerGuard)
@UseInterceptors(ClassSerializerInterceptor)
export class UserApplicationsController {
...
```

where OwnerGuard is defined as follows: 

```
import { Injectable, CanActivate, ExecutionContext } from "@nestjs/common"

@Injectable()
export class OwnerGuard implements CanActivate {
  canActivate(context: ExecutionContext): boolean {
    const request = context.switchToHttp().getRequest()
    return request.user.id === request.params["userId"]
  }
}
```

Maybe it's a good moment to think about https://github.com/bloom-housing/bloom/issues/315  ?

pbn4 commented:
Ready for review

pbn4 commented:
@software-project I took your listings seeding code and wrapped into a service so that seeder now allows using dependency injection and services (it was useful when creating a user with hashed password)

jaredcwhite commented on backend/core/src/listings/listings.service.ts:
```diff
@@ -1,23 +1,39 @@
+  constructor(@InjectRepository(Listing) private readonly repo: Repository<Listing>) {}
+
+  public async list(options?: ListingsListQueryParams): Promise<ListingsListResponse> {
+    let listings = await this.repo.find({ relations: ["units", "attachments", "preferences"] })
```
Whoops, was this a regression? (changing from `Listing.find`?)

pbn4 replied:
Fixed

pbn4 commented:
@jaredcwhite their purpose is described in here https://github.com/bloom-housing/bloom/issues/130 . 

>  One seems to be for public access of data,

What do you mean by public? One must have an admin role to access `/applications`. This is created specifically for superusers querying over applications table. 

> Can we simplify the API endpoint down to one and add logic there to allow or deny requests based on if auth is required?

Answer to this question depends on the answer to the previous one.

bencpeters commented on backend/core/src/listings/listings.module.ts:
```diff
@@ -0,0 +1,13 @@
+import { Listing } from "../entity/listing.entity"
+
+@Module({
+  imports: [TypeOrmModule.forFeature([Listing])],
```
Maybe list the other related entities here (`unit`, `attachment`?) rather than at the root level?

pbn4 replied:
Done

pbn4 replied:
Done

bencpeters commented on backend/core/src/seeder/listings-seeder/listings-seeder.service.spec.ts:
```diff
@@ -0,0 +1,18 @@
+  });
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
```
Do we want this placeholder test here? I saw you deleted the other ones...

pbn4 replied:
Done

pbn4 replied:
Removed.

bencpeters commented on backend/core/src/seeder/listings-seeder/listings-seeder.service.ts:
```diff
@@ -0,0 +1,50 @@
+    for await (const listing of listings) {
+      const l = new Listing()
+      const entities = {}
+      for await (const key of Object.keys(listing)) {
```
You can use `for (const <property> in <object>)` to iterate through objects.

pbn4 replied:
That code I was just copying and wrapping into a service, but OK I'll change that.

pbn4 replied:
Done

bencpeters commented on backend/core/src/user-applications/user-applications.controller.ts:
```diff
@@ -0,0 +1,63 @@
+
+@Controller("user")
+@UseGuards(JwtAuthGuard)
+@ApiBearerAuth()
```
Isn't this duplicating the JWT logic from the Auth guard? It checks for a `Bearer token` too...

pbn4 replied:
It is not. It only adds metadata for swagger generation. See 
https://github.com/nestjs/swagger/blob/master/lib/decorators/api-bearer.decorator.ts and https://github.com/nestjs/swagger/blob/master/lib/decorators/api-security.decorator.ts

bencpeters commented:
> @bencpeters Are you the author? Is this related to our monorepo setup? I have a side project using NestJS and I did not hit that anywhere.

@pbn4 So, the issue is that we need access to `ormconfig.ts` in two different places. One is imported using TS-style imports into the application loader, the other is imported by the `typeorm` cli utility for configuration. The type ORM CLI appears to require Common JS style exports with the export being an object with keys and values:

```javascript
export = {
  type,
  url,
  ...
}
```

But for the application, we want to be able to import all of the exports as an object that we can spread:

```javascript
    TypeOrmModule.forRoot({
      ...dbOptions,
      autoLoadEntities: true,
    }),
```

If we didn't have to support common JS too, we would just do this:

```typescript
// Export definition (in ormconfig.ts)
const options = { ... };
export default options;

// Import
import dbOptions from '../ormConfig'
```

But the CLI doesn't appear to be able to pick that up correctly.

jaredcwhite commented:
> @jaredcwhite their purpose is described in #130 .
> 
> > One seems to be for public access of data,
> 
> What do you mean by public? One must have an admin role to access `/applications`. This is created specifically for superusers querying over applications table.
> 
> > Can we simplify the API endpoint down to one and add logic there to allow or deny requests based on if auth is required?
> 
> Answer to this question depends on the answer to the previous one.

I guess what I'm driving at is I would expect a single API endpoint (`/applications`) to deal with the data model "applications". What that endpoint is allowed do (aka when querying or saving data) depends on the request auth. Let me know if we should hop on a chat to discuss further.

pbn4 commented on backend/core/ormconfig.test.ts:
```diff
@@ -0,0 +1,33 @@
+  port: 5432,
+  database: "bloom_test",
+  synchronize: true,
+  logging: false,
```
Done

pbn4 commented on backend/core/package.json:
```diff
@@ -17,7 +17,9 @@
     "typeorm": "ts-node ./node_modules/.bin/typeorm",
+    "typeorm-test": "ts-node ./node_modules/.bin/typeorm --config ./ormconfig.test.ts",
     "herokusetup": "node heroku.setup.js",
+    "test:seed": "ts-node src/seed.ts --test",
```
Yup it actually makes sense once I looked at other scripts in package.json, those starting with `test:`  are actually testing something. :)

pbn4 commented on backend/core/src/app.module.ts:
```diff
@@ -1,28 +1,31 @@
       autoLoadEntities: true,
     }),
-    TypeOrmModule.forFeature([Listing, Attachment, Preference, Unit]),
+    TypeOrmModule.forFeature([Listing, Attachment, Preference, Unit, Application]),
```
I agree. The same applies also for listing related entities registered here, since I created a module for them too.

pbn4 commented on backend/core/src/auth/jwt.strategy.ts:
```diff
@@ -17,7 +17,7 @@ export class JwtStrategy extends PassportStrategy(Strategy) {
       jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
       passReqToCallback: true,
-      ignoreExpiration: false,
+      ignoreExpiration: true,
```
That was my dev setup that accidentially got into the PR

pbn4 commented on backend/core/src/applications/applications.service.ts:
```diff
@@ -0,0 +1,10 @@
+@Injectable()
+export class ApplicationsService {
+  async find(userId?: string, listingId?: string): Promise<Application[]> {
+    return await Application.find(pickBy({ userId, listingId }, identity))
```
Done

pbn4 replied:
Done

pbn4 commented on backend/core/src/auth/login.dto.ts:
```diff
@@ -0,0 +1,14 @@
+  @IsString()
+  password: string
+}
+
+export class LoginResponseDto {
+  @IsEmail()
```
The fact that it worked means it was not validated after all. I fixed that but the quesiton is whether we want to validate our responses.

pbn4 commented on backend/core/src/entity/application.entity.ts:
```diff
@@ -0,0 +1,43 @@
+  @ManyToOne((type) => User, (user) => user.applications)
+  user: User
+
+  @Column()
+  @IsString()
+  userId: string
```
It allows flat response object structure (one could achieve this with querybuilder and without new column too, but much more complicated), otherwise a response would contain a user object:

```
{
  ...
  user: {
    id: id
  }
},
```
 instead of 

```
{
  ...
  userId: id
}
```

I'm eager to hear comments about this approach.

pbn4 replied:
@bencpeters up

pbn4 commented on backend/core/src/auth/owner.guard.ts:
```diff
@@ -0,0 +1,9 @@
+export class OwnerGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request = context.switchToHttp().getRequest()
+    return request.user.id === request.params["userId"]
```
Good point, I tried. Here is my stackoverflow question https://stackoverflow.com/questions/62712757/how-to-create-explicit-dependencies-between-auth-guards-in-nestjs

pbn4 commented on backend/core/src/auth/roles.guard.ts:
```diff
@@ -0,0 +1,25 @@
+import { Reflector } from "@nestjs/core"
+
+@Injectable()
+export class RolesGuard implements CanActivate {
```
Answered that in question about OwnerGuard

pbn4 commented on backend/core/src/applications/applications.controller.ts:
```diff
@@ -0,0 +1,30 @@
+import { Roles } from "../auth/roles.decorator"
+
+@Controller("applications")
+@UseGuards(JwtAuthGuard)
```
> > @jaredcwhite their purpose is described in #130 .
> > > One seems to be for public access of data,
> > 
> > 
> > What do you mean by public? One must have an admin role to access `/applications`. This is created specifically for superusers querying over applications table.
> > > Can we simplify the API endpoint down to one and add logic there to allow or deny requests based on if auth is required?
> > 
> > 
> > Answer to this question depends on the answer to the previous one.
> 
> I guess what I'm driving at is I would expect a single API endpoint (`/applications`) to deal with the data model "applications". What that endpoint is allowed do (aka when querying or saving data) depends on the request auth. Let me know if we should hop on a chat to discuss further.
> 
> **Update:** sorry, I just noticed your link to #130. Maybe we can discuss with Ben then…I'm not convinced admin routes vs user-specific nested routes makes sense, it seems to make the codebase much more complex.

@jaredcwhite I find API easier to reason about when the endpoint semantics (instead of a query param) tell you that these applications are in a relation to a specific user. I don't have any strong arguments for that and an `/applications` with query params and request auth handling would probably do the work too. I just doubt that it would be less complex than what I proposed. Maybe we can engage more people into the discussion since we are probably going to create few similar endpoints in the future @software-project @bk3c @bencpeters 

BTW: `/applications` is optional and I created it as a convenience. Everything related to applications that we need at the moment is doable using `/users/id/applications`, correct? Superuser interface is not really specified at the moment and I created /applications just as a proposition of how such access to arbitrary application could work.

pbn4 replied:
@bencpeters As discussed elsewhere this must be JWT because roles are required on user object. Do you agree?

pbn4 commented:
> > @jaredcwhite their purpose is described in #130 .
> > > One seems to be for public access of data,
> > 
> > 
> > What do you mean by public? One must have an admin role to access `/applications`. This is created specifically for superusers querying over applications table.
> > > Can we simplify the API endpoint down to one and add logic there to allow or deny requests based on if auth is required?
> > 
> > 
> > Answer to this question depends on the answer to the previous one.
> 
> I guess what I'm driving at is I would expect a single API endpoint (`/applications`) to deal with the data model "applications". What that endpoint is allowed do (aka when querying or saving data) depends on the request auth. Let me know if we should hop on a chat to discuss further.
> 
> **Update:** sorry, I just noticed your link to #130. Maybe we can discuss with Ben then…I'm not convinced admin routes vs user-specific nested routes makes sense, it seems to make the codebase much more complex.

@jaredcwhite I find API easier to reason about when the endpoint semantics (instead of a query param) tell you that these applications are in a relation to a specific user. I don't have any strong arguments for that and an `/applications` with query params and request auth handling would probably do the work too. I just doubt that it would be less complex than what I proposed. Maybe we can engage more people into the discussion since we are probably going to create few similar endpoints in the future @software-project @bk3c @bencpeters 

BTW: `/applications` is optional and I created it as a convenience. Everything related to applications that we need at the moment is doable using `/users/id/applications`, correct? Superuser interface is not really specified at the moment and I created /applications just as a proposition of how such access to arbitrary application could work.

pbn4 commented:
@bencpeters @jaredcwhite I addressed your comments, waiting for replies now. :) 

@bencpeters Weirdly some of the comments were visible here in conversation tab, but not all of them. Some were in 'files changes' tab, so if I missed something please let me know.

pbn4 commented:
@bencpeters Could you elaborate on how we can make explicit dependencies between decorators?

pbn4 commented:
@jaredcwhite I decided to go with what you originally proposed: a single controller and service, not a single service for two controllers. Mainly because my approach had to check path parameter userId against userId that was specified in request body for e.g. POST creating an application (so that I was sure a user if creating an application for himself) and since I was checking the auth data in the service anyway I realised this is not much different than what you proposed and the entire  `/user/:userID` prefix is unnecessary. 

Also, I removed the superuser route for the moment. There is an open issue for how we are going to handle permissions  https://github.com/bloom-housing/bloom/issues/315 so I'm going to bring that back once we have this resolved and actually need a route like that.
